src/transformers/tokenization_utils_base.py

```diff
@@ -1965,24 +1965,55 @@ def convert_added_tokens(obj: Union[AddedToken, Any]):
             # Sort added tokens by index
             added_tok_encoder_sorted = list(sorted(added_tok_encoder.items(), key=lambda x: x[1]))
 
+            # Accumulate added tokens into batches of special/non-special tokens
+            added_tok_batch_is_special = None
+            added_tok_batch_accumulator = []
+
+            def empty_accumulator() -> None:
```

Comment:
Using inner functions with n`onlocal` variables is super dangerous IMO (tricky to detect and fix kind of bugs), we should really try to avoid them if we can.

Is there a reason why:

```python
special_tokens = []
tokens = []
is_last_special = None
for token, index in added_tok_encoder_sorted:
    is_special = token in special_tokens
    if is_special == is_last_special:
        tokens.append(token)
        tokens = []
     elif is_last_special:
        tokenizer.add_tokens(tokens, special_tokens=True)
     else:
         tokenizer.add_tokens(tokens, special_tokens=True)
```

It should achieve the same thing, no ? 

TBH I think an even better fix should exist (but this is definitely good enough once we can get rid of `nonlocal` and inner function.

Reply:
@Narsil Thank you for your review! A better fix would be if we could just accumulate all special and non-special tokens and then call `tokenizer.add_tokens()` twice at the end. However, since we need to need to maintain the order of tokens in the vocab, that's not straightforward. However, unless there is a malicious input, where special tokens and non-special tokens are interleaved, my approach should give the same speed with smaller diff against current code.

Reply:
Your code seems easier to read and shorter. I will incorporate it shortly.

Reply:
@Narsil I simplified the code following your suggestion in 0cd16a8d0.